Implement Nether Portals

xNatsuri opened the pull request:
This PR is built off of #426 
<img width="958" height="453" alt="image" src="https://github.com/user-attachments/assets/bcf5e797-5116-4b65-b3d1-ff55a2fdf579" />
<img width="1485" height="810" alt="image" src="https://github.com/user-attachments/assets/a95ed949-b81f-41d3-91fe-f72e439b8712" />

xNatsuri committed:
feat: implement nether portals

xNatsuri committed:
add obsidian and a nether portal frame block

xNatsuri committed:
changes

TwistedAsylumMC commented on server/world/world.go:
```diff
@@ -257,6 +258,13 @@ func (w *World) setBlock(pos cube.Pos, b Block, opts *SetOpts) {
+	}); ok && !p.Place(pos, w) {
+		// Don't place the block.
+		return
+	}
```
Is there anything that's using this?

xNatsuri replied:
yes flint and steel

didntpot committed:
Merge remote-tracking branch 'upstream/master' into feature/portal

didntpot committed:
Merge branch 'master' into feature/portal

HashimTheArab committed:
Merge branch 'master' into feature/portal

HashimTheArab committed:
Use block-inside events for portal travel

  Portal blocks should not expose collision boxes just to trigger
  interdimensional travel. Route portal contact through the existing
  EntityInside path and let player travel state handle per-tick entry/reset.

  Constraint: Portal blocks are non-colliding, so BBox cannot be the travel trigger.
  Rejected: Scanning nearby portal blocks through Model().BBox | depends on collision
  geometry and duplicates EntityInsider handling.
  Confidence: high
  Scope-risk: moderate
  Directive: Keep portal contact detection in block EntityInside unless Dragonfly adds a
  broader non-collision trigger surface.

HashimTheArab committed:
NeighbourUpdateTick now skips portal rescans for perpendicular horizontal neighbour updates, while still checking vertical updates and same-axis updates so frame breaks are preserved

HashimTheArab committed:
server: tighten Nether portal scan, frame and travel handling

  - Replace string-based portal scan matchers with typed PortalInterior
    and Frame block interfaces (Air, Fire, Obsidian opt in); crying
    obsidian no longer counts as frame and soul fire no longer
    activates a portal.
  - Rewrite scan as a directed rectangular validator with explicit
    width/height bounds, replacing the BFS that accepted any connected
    air/fire region.
  - Add connectedNetherPortal flood-fill so DeactivateNetherPortal
    clears every reachable portal block when a frame is broken,
    avoiding orphan portal blocks left behind by a partial scan.
  - Switch portal travel to a push model: Portal.EntityInside calls
    TravelThroughPortal on the entity; TravelComputer.EnterPortal
    records the contact and StopTravelling only resets the timer when
    the entity leaves.
  - Hoist activation into portal.ActivateNetherPortal and use it from
    flint and steel, fire charge and Fire.Start; remove the racy
    Fire.Place portal hook.
  - Add cube.Axis.Faces() returning the negative and positive Face
    along an Axis, and use it in the portal scan.
  - Drop the Portal model BBox; EntityInside dispatch handles travel
    without needing a collision shape.

HashimTheArab committed:
server: add Nether portal and travel tests, run them in CI

  - Add server/world/portal/nether_portal_test.go covering scan
    acceptance and rejection (axis, frame size limits, missing side
    frames, crying obsidian, soul fire) plus axis-preference,
    ActivateNetherPortal and the FireCharge ignite path.
  - Add server/entity/travel_test.go for TravelComputer.StopTravelling,
    asserting it preserves the timer while the entity is inside a
    portal and resets it once contact ends.
  - Wire `go test ./...` into the pr.yml and push.yml GitHub Actions
    workflows so regressions are caught alongside `make lint`.

HashimTheArab committed:
remove useless docs

HashimTheArab committed:
remove useless docs

HashimTheArab committed:
add light emission to portal block

HashimTheArab committed:
remove Place interface check:
  - No remaining implementer of Place(pos cube.Pos, w *World) bool under server/.
  - Flint and steel does not use that hook now. It activates portals directly in
    item.FlintAndSteel.UseOnBlock via portal.ActivateNetherPortal(tx, s).
  - Fire charge does the same.
  - Fire spread/start also uses portal.ActivateNetherPortal directly.
  - The hook is on World.setBlock, so keeping it would add a dead implicit interface check
    to every block placement.

HashimTheArab committed:
portal/nether.go: simplify and improve code and clarify doc

HashimTheArab committed:
entity: harden TravelComputer and extract portal coord translation

    - Capture the *EntityHandle returned by RemoveEntity and pass it to AddEntity in the destination world; Tx docs warn that the original entity is unusable after removal.
    - Await both source.Exec and destination.Exec so removal completes before the destination tries to add or find the entity, fixing the cross-world race.
    - Bail out of travel if RemoveEntity returns nil and reset travelling.
    - Restructure EnterPortal to compute travelNow under the lock and call travel() after releasing it, removing the unlock-relock dance.
    - Short-circuit EnterPortal when the source world has no distinct portal destination, so non-Nether targets no-op cleanly.
    - Add an instantaneous() helper that tolerates a nil Instantaneous field instead of panicking.
    - Drop the dead `t.inside = false` reassignment in StopTravelling and switch the mutex to sync.Mutex (no read-locked path exists).
    - Unexport Travel to travel; the entry point is EnterPortal and an external caller would skip the timed-out / awaiting-travel checks.
    - Extract translatePortalPosition with godoc covering the 1:8 X/Z scaling and Y clamp to the destination's vertical range, replacing the inline transform that shifted Y by the source's Range().Min().

HashimTheArab committed:
Floor Nether portal scaling for negative Overworld coordinates

HashimTheArab committed:
rename variable

HashimTheArab committed:
fix portal blocks being ignored if it didnt have obsidian below it

Portal lookup distance includes Y, so tall portals must be matched by any active interior block, not only bottom-row blocks with frame below

HashimTheArab committed:
fix: Prevent portal-crossed ender pearls from teleporting owners

HashimTheArab committed:
Tighten Nether portal scan failure semantics

Portal scan helpers should return complete validated portal interiors or no positions at all.

HashimTheArab committed:
add portal teleport method to avoid calling handler

HashimTheArab committed:
Merge branch 'master' into feature/portal

HashimTheArab committed:
feat: allow entities to travel through portals

HashimTheArab committed:
remove test file

HashimTheArab committed:
add offset and neighborface

HashimTheArab committed:
use neighborface

FDUTCH commented on server/entity/ent_portal.go:
```diff
@@ -0,0 +1,74 @@
+	box := e.H().Type().BBox(e).Translate(e.Position()).Grow(-0.0001)
+	low, high := cube.PosFromVec3(box.Min()), cube.PosFromVec3(box.Max())
+
+	for y := low[1]; y <= high[1]; y++ {
```
can be simplified using cube.Range3D

HashimTheArab replied:
good catch, thanks

HashimTheArab committed:
Use shared range iteration for portal scans

Constraint: Preserve portal insider scan bounds while reducing bespoke loop code.
Confidence: high
Scope-risk: narrow
Directive: Keep portal contact checks on the entity tick transaction path.
Tested: go test ./server/entity; go test ./server/block/cube
Not-tested: full repository test suite

HashimTheArab committed:
Merge remote-tracking branch 'upstream/master' into natsuri-feature-portal

# Conflicts:
#	server/entity/projectile.go

didntpot committed:
Merge branch 'master' into feature/portal

cqdetdev committed:
server/world: Optimize portal search with sub-chunk palette filtering (#7)

* server/world: Search sub-chunk palettes instead of every block when finding portals

FindNetherPortal previously called tx.Block on every position in a
256x256 column across the full dimension height (~16M block reads), and
force-generated every missing chunk in the search square as a side
effect.

Replace it with tx.BlocksWithin, which walks the chunks in the radius
and checks each sub-chunk's palette for the portal block before
iterating its storage. Sub-chunks whose palettes contain a portal block
are scanned by comparing raw palette indices, avoiding a runtime ID
lookup per block. Unloaded chunks are read transiently from the
provider without being loaded into the world, and missing chunks are
skipped rather than generated. Portals are validated once instead of
once per portal block found.

No persistent block index is kept (unlike Java's POI system): portal
searches only happen on dimension travel, so re-reading palettes per
search is cheap enough and avoids invalidating an index on every block
change.

At radius 128 a search takes ~85us with every chunk loaded, ~57us in a
mostly ungenerated dimension and ~1.6ms reading cold chunks from disk,
against ~2.4s for the previous per-block scan.

* server/world: Clarify BlocksWithin docs, log provider errors once per search

BlocksWithin now documents that only the primary block layer is searched
and that blocks are matched by block state alone. blocksWithin logs only
the first provider error per search so a systemic failure does not log
once per chunk in the radius.

---------

Co-authored-by: HashimTheArab <hashimssoftware@gmail.com>

HashimTheArab committed:
Merge branch 'master' into feature/portal

HashimTheArab committed:
server/block: Regenerate accessor.go for Portal

HashimTheArab committed:
server/entity: Fold portal travel methods into ent.go

HashimTheArab committed:
server/entity: fix stale transaction use and match vanilla portal travel rules

Re-open the entity from its handle inside the new source transaction in
travelQueued: the wrapper the travel was queued with belonged to a
transaction that had already finished, so despawn handlers and viewers
were passed an entity bound to a closed transaction.

Only players now create a portal at the destination when none is linked;
other entities return to their origin. Non-player entities get the
vanilla 15 second travel cooldown, and the destination portal search
uses the vanilla radius per dimension (128 Overworld, 16 Nether),
which also removes most of the synchronous column loading on the
destination world's transaction queue.

HashimTheArab committed:
server/entity,server/world/portal: test portal travel edge cases

Cover coordinate translation, the delayed travel timer, the travel
cooldown, portal creation gating for non-players and X axis frames.

HashimTheArab committed:
server/player: register portal contact every tick

Portal contact was only registered through Move, which returns early
for a player standing still. StopPortalContact then reset the portal
timer each tick, so a motionless player never finished the four second
wait.

HashimTheArab committed:
server/block,server/world/portal: clean up review nits

Remove the unused Face.Offset, hoist the face slice out of the portal
flood-fill loop, document Obsidian.Frame and register Portal in
alphabetical order.

HashimTheArab committed:
server/player: remove redundant per-tick portal contact check

Portal contact is already registered every tick, including for players
standing still: Tick unconditionally calls checkBlockCollisions, which
runs the EntityInsider checks before StopPortalContact. The extra check
was added on the wrong premise that insiders only run through Move.

HashimTheArab committed:
server/world: add GameMode.InstantPortalTravel

Instant portal travel was hardcoded to the creative game mode, leaving
custom game modes unable to opt in. Creative and spectator skip the
wait, matching the invulnerable player abilities in vanilla.

HashimTheArab committed:
server/entity,server/world/portal: consolidate portal tests

Fold the X axis activation test into TestActivateNetherPortal and drop
a timer reset assertion already covered by the StopPortalContact test.

HashimTheArab committed:
server/world,server/player: exclude spectators from portal travel

Spectators are not affected by nether portals in vanilla. Portal travel
is skipped for game modes without collision, and spectator no longer
reports instant portal travel.

HashimTheArab committed:
server/world: search blocks in outward chunk rings

blocksWithin visited chunks in row-major order and FindNetherPortal
consumed the entire radius before choosing the closest portal, loading
up to ~17x17 chunk columns from disk inside the destination world's
transaction. Chunks are now visited in rings growing outward from the
centre, and FindNetherPortal stops as soon as no later ring can beat
the best match, so the common case touches only the first few rings.
Also drop the unused Nether.Bounds and Nether.Deactivate.

HashimTheArab committed:
server/entity: clear arrival latch when portal travel fails

A failed or aborted travel attempt left timedOut set while the entity
sat back inside the source portal, so it could never retry even after
a linked portal was built. The latch is now cleared on failure and the
cooldown alone paces retries.

HashimTheArab committed:
server/world: shorten BlocksWithin doc comments

HashimTheArab committed:
server/world: revert ring-ordered block search

Benchmarks against an on-disk world with all ~17x17 columns saved show
the ring order and early exit gain nothing in practice (8.6ms vs 8.5ms
row-major; the early exit only fires when a further match is yielded,
which a single-portal world never does). Not worth the extra code and
ordering contract. The full 128-radius scan costs ~10ms once per
Overworld arrival; the Nether direction only scans radius 16.

HashimTheArab committed:
server/entity: search 128 blocks for destination portals in both dimensions

Bedrock Edition keeps the full 128-block portal search radius in the Nether
as well as the Overworld; the reduced 16-block Nether radius is Java Edition
behaviour.

HashimTheArab committed:
server/entity: trim portal search radius doc

HashimTheArab committed:
server/entity: stop falling blocks travelling through portals

Falling blocks passing through portals is Java Edition behaviour (added in
1.21); in Bedrock Edition falling blocks do not travel through portals.

HashimTheArab committed:
server/entity: test that falling blocks do not travel through portals

HashimTheArab committed:
server/entity: document per-entity-type portal travel gating

HashimTheArab committed:
server/entity: simplify portal gating doc

HashimTheArab committed:
server/entity: drop contestable lead example from portal gating doc

didntpot reviewed (approved):
LGTM

RestartFU reviewed (approved):
such magnificant code
